src/display/annotation_layer.js

```diff
@@ -38,336 +38,510 @@ var CustomStyle = displayDOMUtils.CustomStyle;
-          break;
-
-        case AnnotationBorderStyleType.BEVELED:
-          warn('Unimplemented border style: beveled');
-          break;
+/**
+ * @typedef {Object} AnnotationElementOptions
+ * @property {Object} data
+ * @property {PDFPage} page
+ * @property {PageViewport} viewport
+ * @property {IPDFLinkService} linkService
+ */
 
-        case AnnotationBorderStyleType.INSET:
-          warn('Unimplemented border style: inset');
-          break;
+/**
+ * @class
+ * @alias AnnotationElementFactory
+ */
+function AnnotationElementFactory() {}
+AnnotationElementFactory.prototype =
+    /** @lends AnnotationElementFactory.prototype */ {
+  /**
+   * @param {AnnotationElementOptions} parameters
+   * @returns {AnnotationElement}
+   */
+  create: function AnnotationElementFactory_create(parameters) {
+    var subtype = parameters.data.annotationType;
+
+    switch (subtype) {
+      case AnnotationType.LINK:
+        return new LinkAnnotationElement(parameters);
 
-        case AnnotationBorderStyleType.UNDERLINE:
-          container.style.borderBottomStyle = 'solid';
-          break;
+      case AnnotationType.TEXT:
+        return new TextAnnotationElement(parameters);
 
-        default:
-          break;
-      }
+      case AnnotationType.WIDGET:
+        return new WidgetAnnotationElement(parameters);
 
-      if (data.color) {
-        container.style.borderColor =
-          Util.makeCssRgb(data.color[0] | 0,
-                          data.color[1] | 0,
-                          data.color[2] | 0);
-      } else {
-        // Transparent (invisible) border, so do not draw it at all.
-        container.style.borderWidth = 0;
-      }
+      default:
+        throw new Error('Unimplemented annotation type "' + subtype + '"');
     }
+  }
+};
 
-    container.style.left = data.rect[0] + 'px';
-    container.style.top = data.rect[1] + 'px';
-
-    container.style.width = width + 'px';
-    container.style.height = height + 'px';
+var AnnotationElement = (function AnnotationElementClosure() {
+  function AnnotationElement(parameters) {
+    this.data = parameters.data;
+    this.page = parameters.page;
+    this.viewport = parameters.viewport;
+    this.linkService = parameters.linkService;
 
-    return container;
+    this.container = this._makeContainer();
   }
 
-  function getHtmlElementForTextWidgetAnnotation(item, page, viewport) {
-    var container = getContainer(item, page, viewport);
-
-    var content = document.createElement('div');
-    content.textContent = item.fieldValue;
-    var textAlignment = item.textAlignment;
-    content.style.textAlign = ['left', 'center', 'right'][textAlignment];
-    content.style.verticalAlign = 'middle';
-    content.style.display = 'table-cell';
+  AnnotationElement.prototype = {
+    /**
+     * Make an empty container for the annotation's HTML element.
+     *
+     * @private
+     * @memberof AnnotationElement
+     */
+    _makeContainer: function AnnotationElement_makeContainer() {
```

Comment:
Nit: I think that calling this `_createContainer` instead feels like better English :)

(Please note: I've only looked through the diff quickly so far, and I still want to go through the code more thoroughly before signing off on it. I'll try to do that during the weekend.)